drivers/entropy: stm32: RNG  configuration on some MCUs

FRASTM opened the pull request:
for the entropy/ RNG of some stm32 mcus (like stm32L5x and stm32U5 and stm32H723) when the RNG peripheral has a 
Health Test Control and COND the sequence to
Two properties are defined by the RNG node : 
```
			health-test-magic = <0x17590abc>;
 			health-test-config = <0x9aae>;
```
According to the ref man, the sequence for writing the RNG_HTCR requires to LL_RNG_EnableCondReset before and LL_RNG_DisableCondReset after writing the RNG_HTCR.
According to the ref Man : "When CONDRST is set to 0 by software its value goes to 0 when the reset process is done.
It takes about 2 AHB clock cycles + 2 RNG clock cycles."
This PR adds a loop to wait for the bit to be read as 0.

Fixes https://github.com/zephyrproject-rtos/zephyr/issues/45866

Signed-off-by: Francois Ramu <francois.ramu@st.com>

FRASTM committed:
drivers: entropy: stm32 driver waits for the CondRst reset

For the stm32 mcu with a CONDRST bit in its RNG_CR
for conditioning Soft Reset, the driver must wait for
the bit to be 0 after disabling. This could take
about 2 AHB clock cycles + 2RNG clock cycles.

Signed-off-by: Francois Ramu <francois.ramu@st.com>

FRASTM commented:
tested with  west build -p auto -b xxx tests/drivers/entropy/api/ 
on  nucleo_u575zi_q and nucleo_l552ze_q
```
*** Booting Zephyr OS build v3.1.0-rc2-12-g74ae95ce7ba9  ***
Running TESTSUITE entropy_api
===================================================================
START - test_entropy_get_entropy
random device is 0x8005b94, name is RNG
  0x8b
  0x47
  0x02                                           
  0x8f                                           
  0x52                                           
  0x2c                                           
  0xf2                                           
  0x01
  0x03
 PASS - test_entropy_get_entropy in 0.10 seconds
===================================================================
TESTSUITE entropy_api succeeded
===================================================================
PROJECT EXECUTION SUCCESSFUL  
```

erwango commented on drivers/entropy/entropy_stm32.c:
```diff
@@ -103,6 +103,9 @@ static void configure_rng(void)
 {
 	RNG_TypeDef *rng = entropy_stm32_rng_data.rng;
 
+#ifdef STM32_CONDRST_SUPPORT
```
How is this flag defined?

FRASTM replied:
the flag is defined by the presence of bit RNG_CR_CONDRST in the RNG CR of the peripheral

erwango replied:
right, I've missed it. Txs

GeorgeCGV reviewed (commented):
Shouldn't lines [106, 108]:
https://github.com/zephyrproject-rtos/zephyr/blob/d8065a5a24a2fa7915fba8d7d0df8a7d265f96fd/drivers/entropy/entropy_stm32.c#L106-L108

and [119, 123]:
https://github.com/zephyrproject-rtos/zephyr/blob/d8065a5a24a2fa7915fba8d7d0df8a7d265f96fd/drivers/entropy/entropy_stm32.c#L119-L123

be moved inside the `DT_INST_NODE_HAS_PROP(0, health_test_config)` check, around `LL_RNG_SetHealthConfig`?

taltenbach commented:
When CONDRST is set, RNG is re-initialized, which adds a latency before entropy to be available. On single core MCUs, `configure_rng()` is done only at entropy driver initialization so this is fine. However, on multi-core MCUs such as STM32H747, `configure_rng()` is invoked for every new entropy generation process and it is probably desirable to avoid useless RNG re-initialization by checking first is RNG_HTCR is already set to the desired value.
Note that currently no STM32 MCUs supporting CONDRST are multi-core, but that might change in the future, leading to performance issues.

FRASTM committed:
dts: rng: stm32 new property for NIST certification

The new property nist-config is introduced to the stm32 mcus
to configure the RNG for NIST SP800-90B certification.

Signed-off-by: Francois Ramu <francois.ramu@st.com>

FRASTM committed:
drivers: entropy: stm32 rng driver configuration for NIST

The driver is writing the DTS nist-config to the RNG periph.
following the validation sequence described by the RefMan.
It depends on the he RNG IP version and is present on some
mcu devices : all with CONDRST bit.
Depends on the RNG CR Autoreset.
Takes the Health test control Register if property is given.

Signed-off-by: Francois Ramu <francois.ramu@st.com>

FRASTM commented:
New version of the driver with the introduction of the `nist-config` property

FRASTM commented:
>

taltenbach commented:
> > On multi-core MCUs such as STM32H747, configure_rng() is invoked for every new entropy generation process and it might be worth checking if RNG_HTCR is already set to the desired value to avoid useless RNG re-initialization.
> 
> In this case, the configure_rng() is done by the acquire_rng(), and could be :
> 
> ```
> static void acquire_rng(void)
> {
> #if defined(CONFIG_SOC_SERIES_STM32WBX) || defined(CONFIG_STM32H7_DUAL_CORE)
> 	/* Lock the RNG to prevent concurrent access */
> 	z_stm32_hsem_lock(CFG_HW_RNG_SEMID, HSEM_LOCK_WAIT_FOREVER);
> 	/* RNG configuration could have been changed by the other core */
> 	if (LL_RNG_GetHealthConfig(rng) != DT_INST_PROP(0, health_test_config)) {
> 		configure_rng();
> 	}
> #endif /* CONFIG_SOC_SERIES_STM32WBX || CONFIG_STM32H7_DUAL_CORE */
> }
> ```

This is not possible because `configure_rng()` does not only write RNG_HTCR and perform NIST-related configuration, it also enables the RNG and RNG IRQ, which could have been disabled by the other core. For example on STM32WB55, the M0 core disable the RNG after use.

GeorgeCGV commented on drivers/entropy/entropy_stm32.c:
```diff
@@ -103,17 +103,46 @@ static void configure_rng(void)
+		(uint32_t)(RNG_CR_CONDRST | DT_INST_PROP(0, nist_config)));
+#endif /* RNG_CR_ARDIS */
+#else /* nist_config */
+	LL_RNG_EnableCondReset(rng);
```
`LL_RNG_EnableCondReset(rng)` is executed only if `#if DT_INST_NODE_HAS_PROP(0, nist_config)`.
Is it then correct to always do `LL_RNG_DisableCondReset(rng);` (_lines [141, 144]_) even if `nist_config` is not present?

mbolivar-nordic reviewed (commented):
No objections from me

erwango commented:
> > > On multi-core MCUs such as STM32H747, configure_rng() is invoked for every new entropy generation process and it might be worth checking if RNG_HTCR is already set to the desired value to avoid useless RNG re-initialization.
> > 
> > 
> > In this case, the configure_rng() is done by the acquire_rng(), and could be :
> > ```
> > static void acquire_rng(void)
> > {
> > #if defined(CONFIG_SOC_SERIES_STM32WBX) || defined(CONFIG_STM32H7_DUAL_CORE)
> > 	/* Lock the RNG to prevent concurrent access */
> > 	z_stm32_hsem_lock(CFG_HW_RNG_SEMID, HSEM_LOCK_WAIT_FOREVER);
> > 	/* RNG configuration could have been changed by the other core */
> > 	if (LL_RNG_GetHealthConfig(rng) != DT_INST_PROP(0, health_test_config)) {
> > 		configure_rng();
> > 	}
> > #endif /* CONFIG_SOC_SERIES_STM32WBX || CONFIG_STM32H7_DUAL_CORE */
> > }
> > ```
> 
> This is not possible because `configure_rng()` does not only write RNG_HTCR and perform NIST-related configuration, it also enables the RNG and RNG IRQ, which could have been disabled by the other core. For example on STM32WB55, the M0 core disables the RNG after use.
> 
> This also assumes the other core won't change NIST-related configuration (in RNG_CR) without also modifying RNG_HTCR, but I guess in any real world scenario all cores in a MCU will use the exact same "NIST configuration", so that may not be actually an issue

@taltenbach so is it fine from this perspective ?

taltenbach commented:
> > > > On multi-core MCUs such as STM32H747, configure_rng() is invoked for every new entropy generation process and it might be worth checking if RNG_HTCR is already set to the desired value to avoid useless RNG re-initialization.
> > > 
> > > 
> > > In this case, the configure_rng() is done by the acquire_rng(), and could be :
> > > ```
> > > static void acquire_rng(void)
> > > {
> > > #if defined(CONFIG_SOC_SERIES_STM32WBX) || defined(CONFIG_STM32H7_DUAL_CORE)
> > > 	/* Lock the RNG to prevent concurrent access */
> > > 	z_stm32_hsem_lock(CFG_HW_RNG_SEMID, HSEM_LOCK_WAIT_FOREVER);
> > > 	/* RNG configuration could have been changed by the other core */
> > > 	if (LL_RNG_GetHealthConfig(rng) != DT_INST_PROP(0, health_test_config)) {
> > > 		configure_rng();
> > > 	}
> > > #endif /* CONFIG_SOC_SERIES_STM32WBX || CONFIG_STM32H7_DUAL_CORE */
> > > }
> > > ```
> > 
> > 
> > This is not possible because `configure_rng()` does not only write RNG_HTCR and perform NIST-related configuration, it also enables the RNG and RNG IRQ, which could have been disabled by the other core. For example on STM32WB55, the M0 core disables the RNG after use.
> > This also assumes the other core won't change NIST-related configuration (in RNG_CR) without also modifying RNG_HTCR, but I guess in any real world scenario all cores in a MCU will use the exact same "NIST configuration", so that may not be actually an issue
> 
> @taltenbach so is it fine from this perspective ?



> > > > On multi-core MCUs such as STM32H747, configure_rng() is invoked for every new entropy generation process and it might be worth checking if RNG_HTCR is already set to the desired value to avoid useless RNG re-initialization.
> > > 
> > > 
> > > In this case, the configure_rng() is done by the acquire_rng(), and could be :
> > > ```
> > > static void acquire_rng(void)
> > > {
> > > #if defined(CONFIG_SOC_SERIES_STM32WBX) || defined(CONFIG_STM32H7_DUAL_CORE)
> > > 	/* Lock the RNG to prevent concurrent access */
> > > 	z_stm32_hsem_lock(CFG_HW_RNG_SEMID, HSEM_LOCK_WAIT_FOREVER);
> > > 	/* RNG configuration could have been changed by the other core */
> > > 	if (LL_RNG_GetHealthConfig(rng) != DT_INST_PROP(0, health_test_config)) {
> > > 		configure_rng();
> > > 	}
> > > #endif /* CONFIG_SOC_SERIES_STM32WBX || CONFIG_STM32H7_DUAL_CORE */
> > > }
> > > ```
> > 
> > 
> > This is not possible because `configure_rng()` does not only write RNG_HTCR and perform NIST-related configuration, it also enables the RNG and RNG IRQ, which could have been disabled by the other core. For example on STM32WB55, the M0 core disables the RNG after use.
> > This also assumes the other core won't change NIST-related configuration (in RNG_CR) without also modifying RNG_HTCR, but I guess in any real world scenario all cores in a MCU will use the exact same "NIST configuration", so that may not be actually an issue
> 
> @taltenbach so is it fine from this perspective ?

It is not in the sense the RNG is reconfigured even if the current configuration was already the proper one. A simple check in `configure_rng()` of the current `RNG_CR` and `RNG_HTCR` values would be sufficient to avoid that. But, as I said, since at present all STM32 MCUs supporting CONDRST are single-core, meaning that `configure_rng()` is only called once at startup, that is not a real issue at the moment, but might be in the future.

FRASTM commented:
> > > > > On multi-core MCUs such as STM32H747, configure_rng() is invoked for every new entropy generation process and it might be worth checking if RNG_HTCR is already set to the desired value to avoid useless RNG re-initialization.
> > > > 
> > > > 
> > > > In this case, the configure_rng() is done by the acquire_rng(), and could be :
> > > > ```
> > > > static void acquire_rng(void)
> > > > {
> > > > #if defined(CONFIG_SOC_SERIES_STM32WBX) || defined(CONFIG_STM32H7_DUAL_CORE)
> > > > 	/* Lock the RNG to prevent concurrent access */
> > > > 	z_stm32_hsem_lock(CFG_HW_RNG_SEMID, HSEM_LOCK_WAIT_FOREVER);
> > > > 	/* RNG configuration could have been changed by the other core */
> > > > 	if (LL_RNG_GetHealthConfig(rng) != DT_INST_PROP(0, health_test_config)) {
> > > > 		configure_rng();
> > > > 	}
> > > > #endif /* CONFIG_SOC_SERIES_STM32WBX || CONFIG_STM32H7_DUAL_CORE */
> > > > }
> > > > ```
> > > 
> > > 
> > > This is not possible because `configure_rng()` does not only write RNG_HTCR and perform NIST-related configuration, it also enables the RNG and RNG IRQ, which could have been disabled by the other core. For example on STM32WB55, the M0 core disables the RNG after use.
> > > This also assumes the other core won't change NIST-related configuration (in RNG_CR) without also modifying RNG_HTCR, but I guess in any real world scenario all cores in a MCU will use the exact same "NIST configuration", so that may not be actually an issue
> > 
> > 
> > @taltenbach so is it fine from this perspective ?
> 
> It is not in the sense the RNG is reconfigured even if the current configuration was already the proper one. A simple check in `configure_rng()` of the current `RNG_CR` and `RNG_HTCR` values would be sufficient to avoid that. But, as I said, since at present all STM32 MCUs supporting CONDRST are single-core, meaning that `configure_rng()` is only called once at startup, so that is not a real issue at the moment, but might be in the future.

 Do you expect acquire_rng() to check the LL_RNG_GetHealthConfig before going to configure_rng() as follows: 
```
static void acquire_rng(void)
{
#if defined(CONFIG_SOC_SERIES_STM32WBX) || defined(CONFIG_STM32H7_DUAL_CORE)
	/* Lock the RNG to prevent concurrent access */
	z_stm32_hsem_lock(CFG_HW_RNG_SEMID, HSEM_LOCK_WAIT_FOREVER);
	/* RNG configuration could have been changed by the other core */
	if (LL_RNG_GetHealthConfig(rng) != DT_INST_PROP(0, health_test_config)) {
		configure_rng();
	}
#endif /* CONFIG_SOC_SERIES_STM32WBX || CONFIG_STM32H7_DUAL_CORE */
}
```

taltenbach commented:
For the reasons exposed above, this is not possible in particular because there is some code in `configure_rng()` that must be executed even if the value of RNG_HTCR is correct. I was thinking of something like the following. Note this is just a quick draft to clarify my idea, I didn't actually test the code.
```
static void configure_rng(void)
{
	RNG_TypeDef *rng = entropy_stm32_rng_data.rng;

#ifdef STM32_CONDRST_SUPPORT
	uint32_t desired_nist_cfg = DT_INST_PROP_OR(0, nist_config, 0U);
	uint32_t desired_htcr = DT_INST_PROP_OR(0, health_test_config, 0U);
	uint32_t cur_nist_cfg = 0U;
	uint32_t cur_htcr = 0U;

#if DT_INST_NODE_HAS_PROP(0, nist_config)
#if defined(RNG_CR_ARDIS)
	cur_nist_cfg = rng->CR & (RNG_CR_NISTC | RNG_CR_CLKDIV | RNG_CR_RNG_CONFIG1 |
				  RNG_CR_RNG_CONFIG2 | RNG_CR_RNG_CONFIG3 | RNG_CR_ARDIS);
#else /* RNG_CR_ARDIS */
	cur_nist_cfg = rng->CR & (RNG_CR_NISTC | RNG_CR_CLKDIV | RNG_CR_RNG_CONFIG1 |
				  RNG_CR_RNG_CONFIG2 | RNG_CR_RNG_CONFIG3);
#endif /* RNG_CR_ARDIS */
#endif /* nist_config */

#if DT_INST_NODE_HAS_PROP(0, health_test_config)
	cur_htcr = LL_RNG_GetHealthConfig(rng);
#endif /* health_test_config */

	if (cur_nist_cfg != desired_nist_cfg || cur_htcr != desired_htcr) {
		rng->CR = (rng->CR & ~cur_nist_cfg) | (desired_nist_cfg | RNG_CR_CONDRST);

#if DT_INST_NODE_HAS_PROP(0, health_test_config)
#if DT_INST_NODE_HAS_PROP(0, health_test_magic)
		LL_RNG_SetHealthConfig(rng, DT_INST_PROP(0, health_test_magic));
#endif /* health_test_magic */
		LL_RNG_SetHealthConfig(rng, desired_htcr);
#endif /* health_test_config */

		LL_RNG_DisableCondReset(rng);
		while (LL_RNG_IsEnabledCondReset(rng)) {
		}
	}
#endif /* STM32_CONDRST_SUPPORT */

	LL_RNG_Enabled(rng);
	LL_RNG_EnableIT(rng);
}
```

FRASTM commented:
Put the sequence to write the RNG_HTCR  in the configure_rng() as exposed above
Tested with a `health-test-magic = <0x00009aae>; `property in the overlay of the stm32u575 nucleo board.

FRASTM committed:
dts: arm: stm32 entropy with nist configuration for the RNG

The nist-config and health-test-magic properties are added
to the dtsi file, in the RNG node for the stm32 devices
that supports the RNG entropy source validation.
The health-test-magic nb exists on stm32l552 stm32h723/7a3.

Signed-off-by: Francois Ramu <francois.ramu@st.com>

FRASTM committed:
boards: arm: stm32u575 nucleo board enables RNG entropy generator

This commit enables the RNG node just on the nucleo_u575zi_q
board like the b_u585i_iot02a.

Signed-off-by: Francois Ramu <francois.ramu@st.com>

FRASTM commented:
>

FRASTM committed:
boards: arm: stm32h7a3 nucleo board enables RNG entropy generator

This commit enables the RNG node just on the nucleo_h7a3zi_q
board like the nucleo_h723zg.

Signed-off-by: Francois Ramu <francois.ramu@st.com>

taltenbach commented on dts/arm/st/l5/stm32l5.dtsi:
```diff
@@ -413,6 +413,8 @@
 			health-test-config = <0xa2b3>;
 			status = "disabled";
 			label = "RNG";
+			health-test-magic = <0xf00d00>;
```
From the value I guess that should be `nist-config` instead of `health-test-magic`

taltenbach commented on dts/arm/st/l5/stm32l5.dtsi:
```diff
@@ -413,6 +413,8 @@
 			status = "disabled";
 			label = "RNG";
+			health-test-magic = <0xf00d00>;
+			health-test-config = <0xa2b3>;
```
Duplicate

taltenbach commented on dts/arm/st/h7/stm32h7a3.dtsi:
```diff
@@ -47,6 +47,11 @@
+
+		rng: rng@48021800 {
+			nist-config = <0xf00d00>;
+			health-test-magic = <0x72ac>;
```
The value of `health-test-magic` should be `0x17590abc`, `0x72ac` is the value of `health-test-config`, which seem to be missing

FRASTM commented on dts/arm/st/u5/stm32u5.dtsi:
```diff
@@ -388,6 +388,8 @@
 			status = "disabled";
 			label = "RNG";
+			nist-config = <0xf60d00>;
+			health-test-magic = <0x9aae>;
```
but, The RM gives a RNG config table: 
[stm32u5_RNG.pdf](https://github.com/zephyrproject-rtos/zephyr/files/9081745/stm32u5_RNG.pdf)

FRASTM commented:
The register magic number ` health-test-magic = <0x17590abc> `   is given by the RM of the stm32h7a3/b3 and the stm32l5x 
"When writing this register magic number 0x17590ABC must be written immediately before the indicated value"

There is no register magic number given for the stm32u5x.

taltenbach commented on dts/arm/st/h7/stm32h7a3.dtsi:
```diff
@@ -47,6 +47,11 @@
+
+		rng: rng@48021800 {
+			nist-config = <0xf00d00>;
+			health-test-magic = <0x17590abc>;
```
`health-test-config` value still seem to be missing

taltenbach commented on dts/arm/st/l5/stm32l5.dtsi:
```diff
@@ -414,6 +414,8 @@
 			status = "disabled";
 			label = "RNG";
+			nist-config = <0xf00d00>;
+			health-test-magic = <0x17590abc>;
```
Duplicate: `health-test-magic` already set at line 413

FRASTM commented:
rebase on bfb65b054cd8cdad4169be2fd6b79b200d22d593
add <nist-config>    <health-test-magic> <health-test-config>  properties on stm32u5 stm32l5 and stm32h7a3 on RNG node

fabiobaltieri commented:
@FRASTM can you rebase this please?

carlescufi commented:
@FRASTM please rebase

FRASTM commented:
- rebase on 5897de1e535816247e309348cbe29760361a28b8
- fix conflict

erwango commented:
@ceolin would you mind re-approve following rebase ?

FRASTM commented:
rebase on f3017aa787e75c77bb7deb9f4ae45d4b4e93fbc3
